### ISSUE TITLE
Support llvm-5

### ISSUE THREAD

**Issue by billtohara** [Morgan-Stanley/hobbes#69]
Building hobbes currently fails when using llvm-5.0 (latest stable release from September).

It appears various parts of the code check for llvm versions up to and including 4.x. Treating 5.0 as 4.x gets the code to build, but the resulting binary does not work. "hobbes-test" and "hi" both fail with an empty "LLVM ERROR". (No other message is printed).

Not being a c++ programmer and not understanding llvm means I have limited ability to patch this. Is support planned for llvm 5.0?

**Comment by kthielen** [comment]
Hi Bill,

We are considering a variety of things re: LLVM.  Our use of LLVM is actually fairly limited currently.  I haven't looked at the 5.0 upgrade but my guess is that it's probably pretty easy.

Is there a particular feature in 5.0 that you're interested in?  Given that you're not a C++ programmer, asking for support for a particular LLVM version seems like an unusual request.  :)

**Comment by billtohara** [comment]
Thanks for the awesomely fast response :-)

I'm trying to build hobbes for internal experiments by some other folks here. I own the build but not the usage. We internally use llvm-5 already because it supports some backend features optimized for recent x86 processors that we've deployed. (My understanding is that we cannot use an older llvm as it doesn't support instruction extensions and is missing optimizations for the new platform).

Is there anything I can do to get a better error message from llvm for you?

**Comment by kthielen** [comment]
Hi Bill,

Not being a C++ programmer, it'd be hard for you to get more details I think.  If hobbes compiled with LLVM 5 then it's probably a small change that's needed.

I'm a little tied up with internal work at the moment, but I (or somebody in the group) should have some time to look into this soon.  In the meantime I will leave this issue open.

**Comment by billtohara** [comment]
Thanks in advance if you do have a chance to look at it. I poked some more this morning and perhaps this will help you or the team.

The symptom of the empty "LLVM ERROR" and hobbes-test exiting on the first test (in Array, Strings) is caused by a change in LLVM-5.0 to the function MCJIT::getSymbolAddress(). The older behavior in 4 appears to be to return an error. In 5.0, it does a "report_fatal_error()" which appears to exit the program.

Looking up the stack, the problem is that jitcc::getSymbolAddress calls ee->getFunctionAddress() and this can exit the program if it fails. I don't know how to fix that and can't see how to check for the function address existing before trying to get it. As a half-assed test, I guarded the ee->getFunctionAddress() with a (slow) call to ee->FindFunctionNamed(). This is not recommended in general due to being slow but I wanted to check whether the function could be found in some "ee". 

Now the test gets further. It appears to try to run some functions but again goes down some error path in LLVM that causes it to unceremoniously exit rather than let you continue. This time the error message is a bit more meaningful (?!):

LLVM ERROR: Program used external function '.intern.str..t6673' which could not be resolved!

I assume this is some generated function but I don't understand what is going on. I can tell you that it checks three different execution engines and fails to find that function.

Hopefully this provides some hints to anyone taking a look .. I'm far beyond my competence at this point!

**Comment by kthielen** [comment]
Wow, thanks Bill!  I think that you've just done most of the detective work on this!  Great job!  You have no business calling yourself incompetent.  ;)

We have had a fun time dealing with all of the JIT changes in LLVM over the last several years.  It sounds like this one comes down to a couple of small adjustments.  :T

**Comment by billtohara** [comment]
Thanks :) One final hint for a real fix. The tests will pass if jitcc::defineGlobal uses ExternalLinkage instead of InternalLinkage when creating GlobalVariables. I have no idea if this rampant pollution of the symbol space has unintended consequences.

Then, when searching for the variables in jitcc:getSymbolAddress(), this test works as intended:

    if (ee->FindFunctionNamed(vn) || ee->FindGlobalVariableNamed(vn)) {
      if (uint64_t faddr = ee->getFunctionAddress(vn)) {
        return (void*)faddr;
      }
    }

But this can't be the real fix. As the LLVM documentation notes for both FindFunctionNamed and FindGlobalVariableNamed: 

    "This is very slow operation and shouldn't be used for general code."

**Comment by kthielen** [comment]
Wow again!  Your highlight of the slowness/inappropriateness of this method is well noted, and we will look for a better way to arrange this code for LLVM 5 (unless you pull another magic trick and beat us to the punch with a PR).

In the meantime, that slowness is presumably just a compile-time issue and perhaps you're able to do some basic testing (with the caveat that all of the LLVM 5 interaction needs improvement)?

**Comment by billtohara** [comment]
Working on this. I'm trying to test a little more to ensure this really works. I'm hitting some problems where it may be that I'm misunderstanding the language or it may be that the llvm-5.0 code doesn't work. Can you confirm this is expected or not? (The segfault should be handled a little better but I don't know if the sequence of expressions should work or not).

    > bob = {x=42, y=53.0, name="bob"}
    > bob.x 
    42
    > bob.x <- 53
    Program received signal SIGSEGV, Segmentation fault.
    0x00007ffff7fe200a in /.t14842 ()

Should that work?

Also, is there some documentation on what the roll and pack/unpack functions do? Thanks!

**Comment by kthielen** [comment]
Hi Bill, that's definitely not expected behavior.  :(

Out of curiosity, if you don't mind, could you try a slightly different but related test?

```
> bob = newPrim()::{x:int,y:double}
> bob.x <- 42

> bob.y <- 3.14159

> bob
{x=42, y=3.14159}
```

I'm curious to know if this works differently for you because although they're both similar in what they do, the first one (your excerpt) will translate to global constant values on the back-end (which LLVM may treat differently in 5.0, perhaps marking the page read-only).

**Comment by billtohara** [comment]
Your test works fine -- I get exactly the same output printed. I'll see if I can figure out why the global constants are uhhh treated as constants ;-)

**Comment by kthielen** [comment]
It might be that this version of LLVM blocks us from writing into constants in this way -- it's kind of an ugly corner.

**Comment by billtohara** [comment]
Yes, that appears to be true. However, a quick experiment in setting the Constant to be a writeable llvm::GlobalVariable doesn't work, as it's never found. I'm guess "hi" is looking in a constant table rather than a variable (the error simply says the global variable can't be found).

I have another reproducible bug though -- I fear I cannot make llvm-5 work due to my limited knowledge of it and c++ (and hobbes!).

Making the mistake of trying to write to x, always causes an allocation of y to fail:

    > x=42
    > x<-53
    stdin:1,1-1: Failed to get reference to global variable: x
    1 x<-53
    
    > y=newPrimZ()::int
    Segmentation fault (core dumped)

Whereas not making the mistake allows the allocation of y to succeed:

    > x=42
    > y=newPrimZ()::int
    > y<-42
    
    > x
    42
    > y
    42

**Comment by kthielen** [comment]
Thank you for this reminder, there's some awkwardness with global definitions and mutation that we should address here.

We don't usually do this kind of thing, though that's not a very good reason to leave the trap door open for somebody to fall through.  :)
